pbxd/pbx/definity.py

```diff
@@ -0,0 +1,411 @@
+                r'Password:',
+            ],
+            timeout=10
+        )
+        if index == 0:  # TIMEOUT
+            self.logger.error('Connection timeout at password:\n{}'.format(self.session.before))
+            raise Exception('Connection timeout at password')
+        elif index == 1:  # EOF
+            self.logger.error('Connection failed with EOF at password:\n{}'.format(self.session.before))
+            error_msg = self.session.before.decode('utf-8').strip().split('\n')[-1]  # last output line
+            raise Exception('Connection failed with EOF at password: {}'.format(error_msg))
+        elif index == 2:  # Password
+            self.logger.debug('Sending pbx_password')
+            self.session.sendline(self.pbx_password)
+
+        self._select_termtype('ossi4')
+
+    def disconnect(self):
+        """
+        Disconnect from the PBX.
+        """
+        self.logger.info('Disonnecting from pbx')
+        if self.session is not None:
+            if self.connected_termtype == 'vt220':
+                self.session.send(b'\x1b[3~')  # VT220 cancel
+                self.session.sendline('logoff')
+            else:
+                self.session.sendline('c logoff')
+                self.session.sendline('t')
+
+            index = self.session.expect(
+                [
+                    pexpect.TIMEOUT,
+                    pexpect.EOF,
+                    'Proceed With Logoff',
+                ]
+            )
+            if index == 0:
+                self.logger.error('Timeout during disconnect:\n{}'.format(self.session.before))
+            elif index == 1:
+                self.logger.error('Connection failed with EOF during disconnect:\n{}'.format(self.session.before))
+            elif index == 2:
+                self.session.sendline('y')
+            self.session.close()
+
+        self.session = None
+        self.connected_termtype = None
+        self.logger.info('Connection closed')
+
+    def reconnect(self):
+        self.logger.warning('Reconnecting...')
+        self.disconnect()
+        self.connect()
+
+    def _select_termtype(self, termtype):
+        """
+        Switch between the ossi4 and vt220 termtypes.
+        """
+        if not self.session.isalive():
+            self.logger.error('dead session: {}'.format(self.session.before))
+            self.reconnect()
+
+        if termtype == self.connected_termtype:
+            return
+
+        if self.connected_termtype == 'vt220':
+            self.session.sendline('newterm')
+        elif self.connected_termtype == 'ossi4':
+            self.session.sendline('c newterm')
+            self.session.sendline('t')
+
+        # Terminal Type (513, 715, 4410, 4425, VT220, NTT, W2KTT, SUNT): [513]
+        index = self.session.expect(
+            [
+                pexpect.TIMEOUT,
+                pexpect.EOF,
+                r'Terminal Type \(.+\): \[.+\]',
+            ]
+        )
+        if index == 0:  # TIMEOUT
```

Comment:
I don't like how this pattern looks but it's what the library documents as how to handle assorted expectations so I guess I can't complain.